weave-js/src/components/UserLink.tsx

```diff
@@ -216,54 +215,59 @@ const UserInner = ({user, includeName, placement}: UserInnerProps) => {
 };
 
 type UserLinkProps = {
-  username: string | null;
+  userId: string | null;
   includeName?: boolean; // Default is to show avatar image only.
   placement?: TooltipProps['placement'];
 };
 
-export const UserLink = ({username, includeName, placement}: UserLinkProps) => {
-  const [user, setUser] = useState<UserResult>(username ? 'load' : 'NA');
-  useEffect(
-    () => {
-      if (user !== 'load') {
-        return;
-      }
-      setUser('loading');
-      apolloClient
-        .query({
-          query: FIND_USER_QUERY as any,
-          variables: {
-            username,
-          },
-        })
-        .then(result => {
-          const {edges} = result.data.users;
-          if (edges.length > 0) {
-            const u = edges[0].node;
-            setUser({
-              ...u,
-              username,
-            });
-          } else {
-            setUser('error');
-          }
-        })
-        .catch(err => {
-          setUser('error');
-        });
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    []
-  );
-  if (user === 'NA') {
+const fetchUser = (userId: string) => {
+  return apolloClient
+    .query({
+      query: FIND_USERS_QUERY as any,
+      variables: {
+        userId,
+      },
+    })
+    .then(result => {
+      return result.data.user as UserInfo;
+    });
+};
+
+const fetchUsers = (userIds: string[]) => {
+  // This is not great, gorilla does not allow multi-user-lookup by id :(
+  return Promise.all(userIds.map(fetchUser));
+};
+
+export const useUsers = (userIds: string[]) => {
+  const memoedUserIds = useDeepMemo(userIds);
+
+  const [users, setUsers] = useState<UserResult>('load');
+  useEffect(() => {
+    setUsers('loading');
+    fetchUsers(memoedUserIds)
+      .then(userRes => {
+        setUsers(userRes);
+      })
+      .catch(err => {
+        setUsers('error');
+      });
+  }, [memoedUserIds]);
+
+  return users;
+};
+
+export const UserLink = ({userId, includeName, placement}: UserLinkProps) => {
+  const users = useUsers(userId ? [userId] : []);
+  if (userId == null) {
     return <NotApplicable />;
   }
-  if (user === 'load' || user === 'loading') {
+  if (users === 'load' || users === 'loading') {
     return <LoadingDots />;
   }
-  if (user === 'error') {
-    return <div>{username}</div>;
+  if (users === 'error') {
+    return <div>{userId}</div>;
```

Comment:
This would show internal ID In the UI as the fallback? I wonder if we should show an error panel instead.

Reply:
fair - i'll replace with a nullish value